ext/openssl/ossl.c

```diff
@@ -604,6 +604,30 @@ static void Init_ossl_locks(void)
 }
 #endif /* !HAVE_OPENSSL_110_THREADING_API */
 
+/*
+ * call-seq:
+ *   OpenSSL.memcmp?("aaa", "aaa")   => true
+ *   OpenSSL.memcmp?("aaa", "bbb")   => false
+ *
+ * Constant time memory comparison
+ */
+static VALUE
+ossl_crypto_memcmp(VALUE dummy, VALUE str1, VALUE str2)
+{
+    const unsigned char *p1 = (const unsigned char *)StringValuePtr(str1);
+    const unsigned char *p2 = (const unsigned char *)StringValuePtr(str2);
+    long len1 = RSTRING_LEN(str1);
+    long len2 = RSTRING_LEN(str2);
+
+    if(len1 != len2)
+    return Qfalse;
+
+    switch (CRYPTO_memcmp(p1, p2, len1)) {
+      case 0:	return Qtrue;
```

Comment:
In order to prevent false positives in `"aaa"` vs. `"aaaa"` -like comparisons, You can return `(len1==len2) ? Qtrue : Qfalse` here instead of just `Qtrue`.  That way the `if` statement above can be eliminated.

Reply:
Do you mean using minimum length for comparison and then doing final length check?

Reply:
Yes.  And I now think I was wrong.

Basically, when you do a timing-safe comparison, one side is "good" and the other side is "bad", like SHA-2 versus user inputs.  It seems C APIs for timing-safe comparison are expecting the length argument to be equal to that of "good" ones, which is normally known before actual comparison.

For us, because this API is going to take only 2 arguments,  I think we have to assume either side being "good" and take its length, like `OpenSSL.memcmp?(good, bad)` or vice versa.  This is fragile.  I can think of several ways to reroute:

- Change the API so that we can take third argument: `OpenSSL.memcmp?(good, bad, length=good.length)`.  This way it is up to the user to properly pass expected length.  Not friendly, though.
- If we are going with the current API: allocate buffers internally, align length of both sides, fill them with the arguments + zero padded, and do the actual comparison.  This might introduce another problem (attacker can feed 1TB junk password).
- Like @bdewater wrote other implementations just early return so it might be okay as-is.
- Not sure about raising exceptions, at the moment.

Reply:
What is the risk of knowing the secret length?

Reply:
- An attacker can find users with the shortest/weakest passwords making them an easier attack target.
- An attacker doesn't have to try secrets with a smaller secret length. This is can be around a 1-2% advantage for random characters (1/number of characters).
See also: https://security.stackexchange.com/a/92238

A better question might be: what's a valid case for comparing different length strings? Shouldn't this method always compare hashes anyway?